tasks/notify.py

```diff
@@ -0,0 +1,293 @@
+        json.dump(job_executions, f)
+    ctx.run(f"{AWS_S3_CP_CMD} {JOB_FAILURES_FILE} {S3_CI_BUCKET_URL}/{JOB_FAILURES_FILE} ", hide="stdout")
+
+
+def retrieve_job_executions(ctx):
+    """
+    Retrieve the stored document in aws s3, or create it
+    """
+    try:
+        ctx.run(f"{AWS_S3_CP_CMD}  {S3_CI_BUCKET_URL}/{JOB_FAILURES_FILE} {JOB_FAILURES_FILE}", hide=True)
+        with open(JOB_FAILURES_FILE) as f:
+            job_executions = json.load(f)
+    except UnexpectedExit as e:
+        if "404" in e.result.stderr:
+            job_executions = create_initial_job_executions()
+        else:
+            raise e
+    return job_executions
+
+
+def create_initial_job_executions():
+    job_executions = {"pipeline_id": 0, "jobs": {}}
+    with open(JOB_FAILURES_FILE, "w") as f:
+        json.dump(job_executions, f)
+    return job_executions
+
+
+def update_statistics(job_executions):
+    # Update statistics and collect consecutive failed jobs
+    alert_jobs = {"consecutive": [], "cumulative": []}
+    failed_jobs = get_failed_jobs(PROJECT_NAME, os.getenv("CI_PIPELINE_ID"))
+    failed_set = {job["name"] for job in failed_jobs.all_failures()}
+    current_set = set(job_executions["jobs"].keys())
+    # Insert data for newly failing jobs
+    new_failed_jobs = failed_set - current_set
+    for job in new_failed_jobs:
+        job_executions["jobs"][job] = {"consecutive_failures": 1, "cumulative_failures": [1]}
+    # Reset information for no-more failing jobs
+    solved_jobs = current_set - failed_set
+    for job in solved_jobs:
+        job_executions["jobs"][job]["consecutive_failures"] = 0
+        job_executions["jobs"][job]["cumulative_failures"].append(0)
+        # Truncate the cumulative failures list
+        if len(job_executions["jobs"][job]["cumulative_failures"]) > CUMULATIVE_LENGTH:
+            job_executions["jobs"][job]["cumulative_failures"] = job_executions["jobs"][job]["cumulative_failures"][1:]
+    # Update information for still failing jobs and save them if they hit the threshold
+    consecutive_failed_jobs = failed_set & current_set
+    for job in consecutive_failed_jobs:
+        job_executions["jobs"][job]["consecutive_failures"] += 1
+        job_executions["jobs"][job]["cumulative_failures"].append(1)
+        # Truncate the cumulative failures list
+        if len(job_executions["jobs"][job]["cumulative_failures"]) > CUMULATIVE_LENGTH:
+            job_executions["jobs"][job]["cumulative_failures"] = job_executions["jobs"][job]["cumulative_failures"][1:]
+        # Save the failed job if it hits the threshold
+        if job_executions["jobs"][job]["consecutive_failures"] == CONSECUTIVE_THRESHOLD:
+            alert_jobs["consecutive"].append(job)
+        if sum(job_executions["jobs"][job]["cumulative_failures"]) == CUMULATIVE_THRESHOLD:
+            alert_jobs["cumulative"].append(job)
+    return alert_jobs, job_executions
+
+
+def send_notification(alert_jobs):
```

Comment:
Nit (doesn't need to be fixed in this PR): it'd be nice to add to the notificaiton:
- links to CI Visibility so we can see the list of failures, and/or a Gitlab link to the last job failure
- the exact number of consecutive / cumulative failures, instead of the threshold.

Reply:
Concerning the second point, I only raise when I'm exactly at the threshold (as I didn't want to send indefinite notifications, just the moment when things are becoming wrong). As such the notifications might flicker in case of flakiness, I thought it was preferrable than constant notifications until the failed job is fixed, wdyt?

Reply:
Besides I was wondering if I should push the consecutive/cumulative failures to ci-visibility and have a datadog monitor on these rather than an ad-hoc slack notification: it would ease the link

Reply:
I think that, in the long run, these numbers should be sent as metrics indeed, that way we could use Datadog monitors for the alerts instead of these manual notifications. But I think we can do these in a second step?